### PR TITLE
Custom ClassLoader for indexing classes

### PR DESCRIPTION
**Pull request by rpau**
@cal101 feel free to review it :)

### CODE REVIEW

**Comment by rpau** [comment]
@cal101 what I have realized is that using ASM (bytecode parser) is probably much faster than using reflection, specially to read the class definition. Now, the critical point is `addType` in `TypesLoaderVisitor`.

https://github.com/EsotericSoftware/reflectasm

**Comment by cal101** [comment]
Reflectasm sounds cool but is reflection really a (measured?) problem?
Or do you think about avoiding class loading via class loser and replace that with reflectasm?
I wonder what that means for access of depended classes.

Since I got some strange results when profiling the class loader related code when I profiled that I used some simple timer based counting System.currentTime*. I found the results more reliable.

Regarding the review I will try to find a slot.

File: src/main/java/org/walkmod/javalang/compiler/types/ClassPathTree.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/ClassPathTree.java]: The add method looks strange to me. I suggest writing a test that shows application of the class. Especially dealing with different depths ok key pathes. 
I would call the keys parameter a keyPath.
I wonder if the outcome of adding
1. A/B/C
2. A/D/E
is correct according to the assumed expected behavior.
All fields seem to be candidates for "final".
Telling the syntax of the key would be nice: absolute file path? Relative file path?

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Make copy of array?

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: I expected this declaration at beginning of class.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Whats the urlclasspath definition for file archive?
In the past we had .zip accepted. Years ago but would check against the urlclassloader.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Throw ioexception instead of assert. See methods javadoc.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Move before outer if to store only without trailing slashes. Strip trailingvslashes on lookup, too if not already done.
File separator is always char and has constants for string and char.

File: src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java]: Here and other places: why not just use for loop instead of explicit iterator?

File: src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java:174]: Do you care for avoiding statics at the moment or do you want deal with them later?
There is some kind of context missing that brackets an execution/run.
I am dreaming of parallel execution here...

**Review by cal101 — CHANGES_REQUESTED** [review]
Check comments. Ignore where too picky... ;-)

**Comment by rpau** [comment]
@cal101 thanks for your comments. Now I will review and fix one by one :)

BTW: about the performance improvement replacing reflection by ASM: https://github.com/rpau/javalang-compiler/pull/86

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: why do you suggest to copy this array?

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Ops it was because the code was initially copied from the pants project

File: src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java:174]: Really good point about the parallel execution here :)

Well, I think that I cannot add more statics for now without overcomplicating the code. Indeed with visualVM now, the hotspot is the `LoadStaticImportsAction` because of the reflection usage there asking for the import class. 

What do you mean with a missing context?

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: URLClasspath belongs to the "misc" classes of the JVM to load resources.
I have excluded zips because I did not about this, but I can include them as well.

File: src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java
Comment (cal101) [reply src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java:174]: If you want to do parallel runs each run needs it's own set of mutable state.
Often ThreadLocals are used but I prefer explicit passing of a kind of "execution context" that stores data global for a run. For single threaded apps the "statics" provide that.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: I suggest checking how urlclassloader handles non-directories and do the same. If they check for suffixes we can do the same. But I expectvsomething different.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Deep mistrust ;-) The array is writable and we are creating an alias. I don't want to check the control flow to make sure all users behave and don't modify. Immutable data types or defensive copy make it possible to keep it private/unmodifyable. Parallel execution plans make that even more important.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: URLConnection urlc = url.openConnection();
            InputStream is = urlc.getInputStream();
            if (urlc instanceof JarURLConnection) {

is what urclassloader does.
I suggest trying to open each non-directory as jar and ignore errors if it's not a jar.

File: src/main/java/org/walkmod/javalang/compiler/types/ClassPathTree.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/ClassPathTree.java]: Fixed in e36e5a883c9cacde332973b540cfc887e8a73c9f

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: I have marked it as final in 5e6fb8984b1ca86bed46c0eaabda4bfb7f06648a
The copy is created by who provides the array ;)

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: f385baae40686db0f1cf6db42cf9730ec4aec7ab

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: a95af1e0961d628eb62289eacc0b2d2c32b6272b

File: src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java]: ab5a24338cb02e90219e7085d497a7feab5ab4d3

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: This is not right at all. The `URLClassLoader` is checking the permissions for the file with the approach that you pointed out. 
http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/java/net/URLClassLoader.java#458

However, here: http://grepcode.com/file/repository.grepcode.com/java/root/jdk/openjdk/6-b14/sun/misc/URLClassPath.java#URLClassPath.JarLoader

Comment (cal101) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: I don't see your point/contradiction.

Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: In the JarLoader is not opening a connection, it is iterating the entries. Looking in detail in the code, I am starting to think that I should maintain and index in `TypesLoaderVisitor` but using the current `URLClassLoader`, which already is indexing the entries. Then, the pros are:
- I do not depend on `sun.misc`package, which is not part of the JDK in Java 10.
- I reduce the `TypesLoader` part because the classloader is not needed anymore for the changes that you see in the next PR

The cons are:
- The indexation happens twice.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: a82571997560adc9c5eb0b672fa162c9755e4cb2

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Nevermind. It not indexes the classpath at the begining. Every request is really expensive in the `URLClassLoader` class

File: src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java:174]: Cal, let's work in the parallel part in a different PR. I do not see how to do it easily with the current logic.

**Comment by rpau** [comment]
@cal101 waiting for your approval ;)

File: src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java
Comment (cal101) [reply src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java:174]: Sure. I don't want to *work* on it now. I just wanted to point out not working too much *against* it.

**Comment by cal101** [comment]
Will look at it after some sleep, garden work, shopping, working and driving. Maybe even some snack inbetween ;-)

File: src/main/java/org/walkmod/javalang/compiler/types/CachedClassLoader.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/CachedClassLoader.java]: Style: wild or no wild?

File: src/main/java/org/walkmod/javalang/compiler/types/CachedClassLoader.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/CachedClassLoader.java]: Collections.unmodifyableMap?

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassLoader.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassLoader.java]: Style: Move to top of class?

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Wildcard?

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Final?

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: File.pathSeparator

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Will that work for java 9?
Should we care now or later?

**Review by cal101 — DISMISSED** [review]
Minor style stuff, see comments.

File: src/main/java/org/walkmod/javalang/compiler/types/CachedClassLoader.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/CachedClassLoader.java]: It has been automatically performed by Intellij. I will removed it.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: Well, I would care if rt is not part of java 9

Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: The internal files rt.jar, tools.jar, and dt.jar have been removed. The content of these files is now stored in a more efficient format in implementation-private files in the lib directory. Class and resource files previously in tools.jar and dt.jar are now always visible via the bootstrap or application class loaders in a JDK image.

😱

Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: I am going to create a Issue, because it is a different discussion.

Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java]: https://github.com/rpau/javalang-compiler/issues/87

**Comment by rpau** [comment]
@cal101 could you review the last changes and then approve if you consider? I think that all your styling comments have been addressed.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java:148]: What I meant was:
```
             String cleanRelPath = stripTrailingFileSeparators();
         if (!index.containsKey(cleanRelPath)) {
￼             index.put(cleanRelPath, delegate);
￼         }
```

And using the same strip-Method for lookup method of index, too.
Seems fishy for me to add multiples here but don't strip on lookup.

File: src/main/java/org/walkmod/javalang/compiler/types/PathTree.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/PathTree.java]: Wild card.

File: src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/TypesLoaderVisitor.java]: Wildcard.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassLoader.java
Comment (cal101) [inline src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassLoader.java]: Can be final.

**Review by cal101 — APPROVED** [review]
Mostly style issues.

File: src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java
Comment (rpau) [reply src/main/java/org/walkmod/javalang/compiler/types/IndexedURLClassPath.java:148]: Theoretically, you could have multiple classes from different jars under the same package and the JDK uses the first that appears in the classpath.

**Pull request merged**
